<a href="https://colab.research.google.com/github/venidaydream/Intro_to_Big_Data/blob/main/Taller1_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Taller 1
1.	Usando el fichero bolsa_2021.csv. Realiza los modelos CAPM de Cie, Enagás y Santander y responde a las siguientes preguntas.

Recuerda que para ello primero tendrás que generar el tipo de interés diario, las rentabilidades de cada acción (e Ibex) y las primeras de riesgo de cada acción e Ibex

a)	¿Cuáles de las acciones son defensivas y cuáles ofensivas?

b)	¿Cumple la constante lo que dice el modelo teórico CAPM?

c)	¿Cuánto del riesgo de la acción explicamos con el riesgo del mercado en cada caso?




# Translation


1. Using the file bolsa_2021.csv. Create the CAPM models of Cie, Enagás and Santander and answer the following questions.

Remember that to do this you will first have to generate the daily interest rate, the returns of each share (and Ibex) and the first risk of each share and Ibex

a) Which of the actions are defensive and which are offensive?



b) Does the constant comply with what the CAPM theoretical model says?



c) How much of the stock's risk do we explain with the market risk in each case?


# Solution

Calculate **Risk Premium**:
Calculate the risk premium for each stock by subtracting the risk-free rate from the stock's returns.

**CAPM Model**:
Apply the CAPM model to estimate the expected return for each stock. The CAPM formula is:

Expected Return=Risk-Free Rate+β×(Market Return−Risk-Free Rate)

a) Defensive or Offensive:
Determine whether the stocks are defensive or offensive based on their beta values. Defensive stocks typically have betas less than 1, while offensive stocks have betas greater than 1.

*beta > 1 - ofensiva*

*beta < 1 - defensiva*

b) Constant in CAPM:
Check whether the constant in the CAPM model (alpha) is close to zero. Theoretically, it should be close to zero.
According to CAPM theory, the intercept should theoretically represent this risk-free rate.

*Constant = Intercept. Intercept tells the average value of omitted variables and noise in the model.*

c) Percentage of Risk Explained:
Calculate the percentage of the stock's risk explained by the market risk using the R-squared value.

*R2 - R2 is the coefficient of determination that tells us that how much percentage variation independent variable can be explained by independent variable.The maximum possible value of R2  can be 1, means the larger the R2  value better the regression.*

*The minimum level of return required by investors occurs when the actual return is the same as the expected return, so that there is no risk at all of the return on the investment being different from the expected return. This minimum level of return is called the ‘risk-free rate of return‘*


Stock_CAPM_return = riskfree + stock_beta * risk premium

In [ ]:
import pandas as pd
import numpy as np
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score
import statsmodels.formula.api as smf

# Load data
data = pd.read_csv('bolsa_2021.csv')
# Drop the last two rows
data = data.iloc[:-3]
print(data)

           Date        Cie         Ibez     Enagas  Santander  interes
0    19/10/2020  17.049999  6860.200195  20.100000   1.584892    0.120
1    20/10/2020  17.020000  6927.299805  20.100000   1.628783    0.145
2    21/10/2020  17.200001  6811.500000  20.100000   1.599075    0.165
3    22/10/2020  17.370001  6796.600098  20.100000   1.617092    0.183
4    23/10/2020  17.530001  6893.399902  20.100000   1.671333    0.170
..          ...        ...          ...        ...        ...      ...
252  12/10/2021  21.860001  8935.900391  19.629999   3.383500    0.441
253  13/10/2021  22.040001  8881.400391  19.385000   3.276000    0.443
254  14/10/2021  22.020000  8925.000000        NaN   3.281000    0.458
255  15/10/2021  22.440001  8997.000000        NaN   3.345500    0.464
256  18/10/2021  22.360001  8970.200195        NaN   3.362000    0.492

[257 rows x 6 columns]


In [ ]:
# Extract daily interest rate from the last column
daily_interest_rate = data['interes']

# Calculate daily returns
data['Daily_Return_Ibex'] = data['Ibez'].pct_change()

data['Daily_Return_Cie'] = data['Cie'].pct_change()
data['Daily_Return_Enagas'] = data['Enagas'].pct_change()
data['Daily_Return_Santander'] = data['Santander'].pct_change()

# Calculate risk premium
data['Risk_Premium_Cie'] = data['Daily_Return_Cie'] - daily_interest_rate #Portfolio_Excess
data['Risk_Premium_Enagas'] = data['Daily_Return_Enagas'] - daily_interest_rate #Portfolio_Excess
data['Risk_Premium_Santander'] = data['Daily_Return_Santander'] - daily_interest_rate #Portfolio_Excess

data['Risk_Premium_Ibex'] = data['Daily_Return_Ibex'] - daily_interest_rate #market excess

data

Date        Cie         Ibez     Enagas  Santander  interes  \
0    19/10/2020  17.049999  6860.200195  20.100000   1.584892    0.120   
1    20/10/2020  17.020000  6927.299805  20.100000   1.628783    0.145   
2    21/10/2020  17.200001  6811.500000  20.100000   1.599075    0.165   
3    22/10/2020  17.370001  6796.600098  20.100000   1.617092    0.183   
4    23/10/2020  17.530001  6893.399902  20.100000   1.671333    0.170   
..          ...        ...          ...        ...        ...      ...   
252  12/10/2021  21.860001  8935.900391  19.629999   3.383500    0.441   
253  13/10/2021  22.040001  8881.400391  19.385000   3.276000    0.443   
254  14/10/2021  22.020000  8925.000000        NaN   3.281000    0.458   
255  15/10/2021  22.440001  8997.000000        NaN   3.345500    0.464   
256  18/10/2021  22.360001  8970.200195        NaN   3.362000    0.492   

     Daily_Return_Ibex  Daily_Return_Cie  Daily_Return_Enagas  \
0                  NaN               NaN                  NaN   
1             0.009781         -0.001759             0.000000   
2            -0.016716          0.010576             0.000000   
3            -0.002187          0.009884             0.000000   
4             0.014242          0.009211             0.000000   
..                 ...               ...                  ...   
252           0.004147         -0.009066             0.006409   
253          -0.006099          0.008234            -0.012481   
254           0.004909         -0.000907             0.000000   
255           0.008067          0.019074             0.000000   
256          -0.002979         -0.003565             0.000000   

     Daily_Return_Santander  Risk_Premium_Cie  Risk_Premium_Enagas  \
0                       NaN               NaN                  NaN   
1                  0.027693         -0.146759            -0.145000   
2                 -0.018239         -0.154424            -0.165000   
3                  0.011267         -0.173116            -0.183000   
4                  0.033542         -0.160789            -0.170000   
..                      ...               ...                  ...   
252                0.010302         -0.450066            -0.434591   
253               -0.031772         -0.434766            -0.455481   
254                0.001526         -0.458907            -0.458000   
255                0.019659         -0.444926            -0.464000   
256                0.004932         -0.495565            -0.492000   

     Risk_Premium_Santander  Risk_Premium_Ibex  
0                       NaN                NaN  
1                 -0.117307          -0.135219  
2                 -0.183239          -0.181716  
3                 -0.171733          -0.185187  
4                 -0.136458          -0.155758  
..                      ...                ...  
252               -0.430698          -0.436853  
253               -0.474772          -0.449099  
254               -0.456474          -0.453091  
255               -0.444341          -0.455933  
256               -0.487068          -0.494979  

[257 rows x 14 columns]

# CIE

In [ ]:
#Linear regression Cie
model_CIE = smf.ols(formula='Risk_Premium_Cie ~ Risk_Premium_Ibex', data=data)
results_CIE = model_CIE.fit()
print(results_CIE.summary())

                            OLS Regression Results                            
Dep. Variable:       Risk_Premium_Cie   R-squared:                       0.989
Model:                            OLS   Adj. R-squared:                  0.989
Method:                 Least Squares   F-statistic:                 2.346e+04
Date:                Sun, 12 Nov 2023   Prob (F-statistic):          3.07e-252
Time:                        15:25:14   Log-Likelihood:                 694.84
No. Observations:                 256   AIC:                            -1386.
Df Residuals:                     254   BIC:                            -1379.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
Intercept             0.0004      0.00

You can think of market beta as a measure of your exposure to the broad stock market. For every 1.0% rise (or fall) in the market, you can expect your portfolio to rise (fall) roughly beta %.

A high adjusted r-squared (close to 1) means that the majority of your portfolio's movements can be explained by the factors in your model.

In [ ]:
beta_CIE = results_CIE.params['Risk_Premium_Ibex'] #systematic risk
beta_CIE #offensive

1.00105243020033

In [ ]:
p_value_CIE = results_CIE.pvalues['Risk_Premium_Ibex']
p_value_CIE #Prob (F-statistic):          3.07e-252 pvalue < 0,05 is statistically significant
#p-value is the probability of obtaining the t statistics at least as contradictory to H0  as calculated from assuming that the null hypothesis is true.

3.071127758139694e-252

In [ ]:
alpha_CIE = results_CIE.params['Intercept'] #specific risk; missing factor; cumple CAPM, statistically 0
alpha_CIE #It aligns with the theoretical expectation that the intercept should represent the risk-free rate, and in this case, it's not significantly different from zero.

0.0003929387134970681

In [ ]:
r2_CIE = results_CIE.rsquared #cuanto % se explica con la cartera de mercado
r2_CIE #The maximum possible value of R2  can be 1, means the larger the R2 value better the regression. Explaining x% of the stock variance

0.9892894913561535

# CIE Results

***a)	¿Cuáles de las acciones son defensivas y cuáles ofensivas?***

beta_CIE = results_CIE.params['Risk_Premium_Ibex'] #systematic risk
beta_CIE #ofensiva


***b)	¿Cumple la constante lo que dice el modelo teórico CAPM?***

alpha_CIE = results_CIE.params['Intercept']

specific risk; missing factor; cumple CAPM, statistically 0. It aligns with the theoretical expectation that the intercept should represent the risk-free rate, and in this case, it's not significantly different from zero.

***c)	¿Cuánto del riesgo de la acción explicamos con el riesgo del mercado en cada caso?***

r2_CIE = results_CIE.rsquared

cuanto % se explica con la cartera de mercado. The maximum possible value of R2  can be 1, means the larger the R2 value better the regression. Explaining x% of the stock variance


**Other stuff:**

The coefficient for "Risk_Premium_Ibex" is 1.0011 = beta. This can be interpreted as, for a one-unit increase in "Risk_Premium_Ibex," the "Risk_Premium_Cie" is expected to increase by approximately 1.0011 units.

# Enagas

In [ ]:
#Linear regression Enagas
model_Enagas = smf.ols(formula='Risk_Premium_Enagas ~ Risk_Premium_Ibex', data=data)
results_Enagas = model_Enagas.fit()
print(results_Enagas.summary())

                             OLS Regression Results                            
Dep. Variable:     Risk_Premium_Enagas   R-squared:                       0.985
Model:                             OLS   Adj. R-squared:                  0.985
Method:                  Least Squares   F-statistic:                 1.646e+04
Date:                 Sun, 12 Nov 2023   Prob (F-statistic):          6.31e-233
Time:                         15:25:14   Log-Likelihood:                 653.41
No. Observations:                  256   AIC:                            -1303.
Df Residuals:                      254   BIC:                            -1296.
Df Model:                            1                                         
Covariance Type:             nonrobust                                         
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
Intercept            -0.0050

In [ ]:
beta_Enagas = results_Enagas.params['Risk_Premium_Ibex']
beta_Enagas #defensive

0.9856485671725078

In [ ]:
alpha_Enagas = results_Enagas.params['Intercept'] #specific risk; missing factor; cumple CAPM, statistically 0
alpha_Enagas #It aligns with the theoretical expectation that the intercept should represent the risk-free rate, and in this case, it's not significantly different from zero.

-0.005003068769692123

# Santander

In [ ]:
#Linear regression Santander
model_Santander = smf.ols(formula='Risk_Premium_Santander ~ Risk_Premium_Ibex', data=data)
results_Santander = model_Santander.fit()
print(results_Santander.summary())

                              OLS Regression Results                              
Dep. Variable:     Risk_Premium_Santander   R-squared:                       0.989
Model:                                OLS   Adj. R-squared:                  0.989
Method:                     Least Squares   F-statistic:                 2.272e+04
Date:                    Sun, 12 Nov 2023   Prob (F-statistic):          1.77e-250
Time:                            15:25:14   Log-Likelihood:                 687.73
No. Observations:                     256   AIC:                            -1371.
Df Residuals:                         254   BIC:                            -1364.
Df Model:                               1                                         
Covariance Type:                nonrobust                                         
                        coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------

In [ ]:
beta_Santander = results_Santander.params['Risk_Premium_Ibex']
beta_Santander #offensive

1.012792626409349

In [ ]:
alpha_Santander = results_Santander.params['Intercept'] #specific risk; missing factor; cumple CAPM, statistically 0
alpha_Santander #It aligns with the theoretical expectation that the intercept should represent the risk-free rate, and in this case, it's not significantly different from zero.

0.005568185996839528